Mathlib/Algebra/Group/TypeTags.lean

```diff
@@ -0,0 +1,481 @@
+#align to_add_div toAdd_div
+
+@[simp]
+theorem ofMul_div [Div α] (x y : α) : ofMul (x / y) = ofMul x - ofMul y :=
+  rfl
+#align of_mul_div ofMul_div
+
+@[simp]
+theorem toMul_sub [Div α] (x y : Additive α) : toMul (x - y) = toMul x / toMul y :=
+  rfl
+#align to_mul_sub toMul_sub
+
+instance Additive.involutiveNeg [InvolutiveInv α] : InvolutiveNeg (Additive α) :=
+  { Additive.neg with neg_neg := @inv_inv α _ }
+
+instance Multiplicative.involutiveInv [InvolutiveNeg α] : InvolutiveInv (Multiplicative α) :=
+  { Multiplicative.inv with inv_inv := @neg_neg α _ }
+
+instance Additive.subNegMonoid [DivInvMonoid α] : SubNegMonoid (Additive α) :=
+  { Additive.neg, Additive.sub, Additive.addMonoid with
+    sub_eq_add_neg := @div_eq_mul_inv α _
+    zsmul := @DivInvMonoid.zpow α _
+    zsmul_zero' := @DivInvMonoid.zpow_zero' α _
+    zsmul_succ' := @DivInvMonoid.zpow_succ' α _
+    zsmul_neg' := @DivInvMonoid.zpow_neg' α _ }
+
+instance Multiplicative.divInvMonoid [SubNegMonoid α] : DivInvMonoid (Multiplicative α) :=
+  { Multiplicative.inv, Multiplicative.div, Multiplicative.monoid with
+    div_eq_mul_inv := @sub_eq_add_neg α _
+    zpow := @SubNegMonoid.zsmul α _
+    zpow_zero' := @SubNegMonoid.zsmul_zero' α _
+    zpow_succ' := @SubNegMonoid.zsmul_succ' α _
+    zpow_neg' := @SubNegMonoid.zsmul_neg' α _ }
+
+instance Additive.subtractionMonoid [DivisionMonoid α] : SubtractionMonoid (Additive α) :=
+  { Additive.subNegMonoid, Additive.involutiveNeg with
+    neg_add_rev := @mul_inv_rev α _
+    neg_eq_of_add := @inv_eq_of_mul_eq_one_right α _ }
+
+instance Multiplicative.divisionMonoid [SubtractionMonoid α] : DivisionMonoid (Multiplicative α) :=
+  { Multiplicative.divInvMonoid, Multiplicative.involutiveInv with
+    mul_inv_rev := @neg_add_rev α _
+    inv_eq_of_mul := @neg_eq_of_add_eq_zero_right α _ }
+
+instance Additive.subtractionCommMonoid [DivisionCommMonoid α] :
+    SubtractionCommMonoid (Additive α) :=
+  { Additive.subtractionMonoid, Additive.addCommSemigroup with }
+
+instance Multiplicative.divisionCommMonoid [SubtractionCommMonoid α] :
+    DivisionCommMonoid (Multiplicative α) :=
+  { Multiplicative.divisionMonoid, Multiplicative.commSemigroup with }
+
+instance Additive.addGroup [Group α] : AddGroup (Additive α) :=
+  { Additive.subNegMonoid with add_left_neg := @mul_left_inv α _ }
+
+instance Multiplicative.group [AddGroup α] : Group (Multiplicative α) :=
+  { Multiplicative.divInvMonoid with mul_left_inv := @add_left_neg α _ }
+
+instance [CommGroup α] : AddCommGroup (Additive α) :=
+  { Additive.addGroup, Additive.addCommMonoid with }
+
+instance [AddCommGroup α] : CommGroup (Multiplicative α) :=
+  { Multiplicative.group, Multiplicative.commMonoid with }
+
+/-- Reinterpret `α →+ β` as `Multiplicative α →* Multiplicative β`. -/
+@[simps]
+def AddMonoidHom.toMultiplicative [AddZeroClass α] [AddZeroClass β] :
+    (α →+ β) ≃ (Multiplicative α →* Multiplicative β) where
+  toFun f := {
+    toFun := fun a => toMul (f (ofMul a))
```

Comment:
In mathlib3 this is `toFun := fun a => ofAdd (f (toAdd a))`. This works out definitionally but has quite confusing types - we have `f: α →+ β` and `a: Multiplicative α` and `toAdd a: Additive (Multiplicative α)` and then apply `f` to that. With the version here we apply `f` to `ofMul a: α` which seems more direct. Happy to revert this in case I misunderstood something. (CC @eric-wieser / leanprover-community/mathlib#13553).

Also the `ext; rfl` proofs work as just `rfl` now which I didn't investigate but would blindly guess is due to eta for structures.

Reply:
I don't think `toAdd` has the type you think it does. `toAdd` and `ofAdd` are about multiplicative not additive.

So this change is incorrect and the original was correct.

Reply:
In general I think it's best not to try to "fix" things like this in the port. If they look wrong, open a mathlib 3 PR since that's more likely to be seen by the people who wrote it. Thanks for the ping though!

Reply:
Oh, you are absolutely right. Thanks! Those are some confusing names :)